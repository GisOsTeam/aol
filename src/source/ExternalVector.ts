import { Vector, IVectorOptions } from './Vector';
import { SourceType, SourceTypeEnum } from './types/sourceType';
import { LayerType, LayerTypeEnum } from './types/layerType';

export interface IExternalVectorOptions extends IVectorOptions {}

export class ExternalVector extends Vector {
  /**
   *
   * @param param0, default useSpatialIndex is false for performance issue
   */
  constructor({ useSpatialIndex = false, ...options }: IExternalVectorOptions) {
    super({ ...options, useSpatialIndex });
    this.options = options;
    if (this.options.snapshotable != false) {
      this.options.snapshotable = true;
    }
    if (this.options.listable != false) {
      this.options.listable = true;
    }
    if (this.options.removable != false) {
      this.options.removable = true;
    }
  }

  public getSourceType(): SourceType {
    return SourceTypeEnum.ExternalVector;
  }

  public getSourceOptions(): IExternalVectorOptions {
    return this.options;
  }

  public setSourceOptions(options: IExternalVectorOptions): void {
    this.options = { ...options };
  }

  public getLayerType(): LayerType {
    return LayerTypeEnum.Vector;
  }

  public isSnapshotable(): boolean {
    return this.options.snapshotable;
  }

  public isListable(): boolean {
    return this.options.listable;
  }

  public isRemovable(): boolean {
    return this.options.removable;
  }
}
