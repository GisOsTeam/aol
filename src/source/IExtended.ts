import OlMap from 'ol/Map';
import Source, { Options } from 'ol/source/Source';
import Feature from 'ol/Feature';
import Geometry from 'ol/geom/Geometry';
import Projection from 'ol/proj/Projection';
import { SourceType } from './types/sourceType';
import { LayerType } from './types/layerType';
import { AndPre, IPredicate, Or } from '../filter/predicate';
import { FilterBuilder } from '../filter/FilterBuilder';

export interface ISnapshotOptions extends Options {
  snapshotable?: boolean;
  listable?: boolean;
  removable?: boolean;
}

export interface ISnapshotSource extends Source {
  getSourceType(): SourceType;

  getSourceOptions(): ISnapshotOptions;

  setSourceOptions(options: ISnapshotOptions): void;

  getLayerType(): LayerType;

  isSnapshotable(): boolean; // Can save snapshot
  isListable(): boolean; // Visible by others tools
  isRemovable(): boolean; // Removable from map by others tools
}

export interface IInitSource extends ISnapshotSource {
  init(): Promise<void>;
}

export interface IQuerySource extends ISnapshotSource {
  query(identifyRequest: IGisRequest, onlyVisible?: boolean): Promise<IQueryResponse>;

  retrieveFeature(id: number | string, featureProjection: Projection): Promise<Feature>;
}

export interface ILayerLegend {
  /**
   * Label optinal
   */
  label?: string;
  /**
   * Image element
   */
  image: HTMLImageElement;
  /**
   * Image data in Base64 ou url
   */
  srcImage: string;
  /**
   * Height
   */
  height: number;
  /**
   * Width
   */
  width: number;
}

export type ILegendRecord = Record<number | string, ILayerLegend[]>;

export interface ILegendSource {
  fetchLegend(): Promise<ILegendRecord>;
}

export interface IExtended extends IInitSource, IQuerySource, ILegendSource { }

export type IGisRequest = IQueryRequest | IIdentifyRequest;

export interface IAbstractRequest<T extends string> {
  olMap: OlMap;
  queryType: T;
  filters?: IPredicate;
  limit?: number;
  returnGeometry?: boolean;
  srId?: string;
}

export interface IQueryRequest extends IAbstractRequest<'query'> {
  geometry?: Geometry;
  geometryProjection?: Projection;
  geometryPrecision?: number;
}

export interface IIdentifyRequest extends IAbstractRequest<'identify'> {
  geometry: Geometry;
  geometryPrecision?: number;
  geometryProjection: Projection;
  identifyTolerance?: number; // in pixel
  layersPrefix?: LayersPrefix;
  outFields?: string;
  returnFieldName?: boolean;
}

export interface IQueryResponse {
  request: IGisRequest;
  featureTypeResponses: IQueryFeatureTypeResponse[];
}

export interface IQueryFeatureTypeResponse {
  type: IFeatureType<any>;
  features: Feature[];
  source: IQuerySource;
}

export interface IAttribute {
  key: string;
  type?: 'Oid' | 'Boolean' | 'Number' | 'String' | 'Date' | 'Geometry' | 'Unknown';
  name?: string;
}

export interface IFeatureType<IDT extends number | string> {
  id: IDT;
  queryId?: IDT;
  hide?: boolean;
  name?: string;
  identifierAttribute?: IAttribute;
  attributes?: IAttribute[];
  readonly initialPredicate?: IPredicate;

  joinWithOr?: boolean;

  predicateWrapper?: FeatureTypePredicateWrapper;
}

export class FeatureTypePredicateWrapper {
  public joinWithOr: boolean;

  private _initialPredicate: IPredicate;
  private _additionalPredicates: Record<string, IPredicate> = {};

  public constructor(initialPredicate?: IPredicate, joinWithOr = true) {
    this._initialPredicate = initialPredicate;
    this.joinWithOr = joinWithOr;
  }

  public addPredicate(predicate: IPredicate): void {
    this._additionalPredicates[predicate.hashCode()] = predicate;
  }

  public removePredicate(hash: string) {
    delete this._additionalPredicates[hash];
  }

  get predicate(): IPredicate {
    return this.filterBuilder.predicate;
  }

  get filterBuilder(): FilterBuilder {
    const predicates: IPredicate[] = Object.values(this._additionalPredicates);
    if (predicates.length > 0) {
      let filterBuilder: FilterBuilder;
      for (const predicate of predicates) {
        if(!filterBuilder) {
          filterBuilder = new FilterBuilder(predicate);
        } else {
          if (this.joinWithOr) {
            filterBuilder.or(predicate);
          } else {
            filterBuilder.and(predicate);
          }
        }
      }
      return filterBuilder.and(this._initialPredicate);
    }
    return new FilterBuilder(this._initialPredicate);
  }

}
export type LayersPrefix = LayersPrefixEnum.ALL | LayersPrefixEnum.TOP | LayersPrefixEnum.VISIBLE;
export enum LayersPrefixEnum {
  ALL = 'all',
  TOP = 'top',
  VISIBLE = 'visible',
}
