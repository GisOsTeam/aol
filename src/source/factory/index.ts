/**
 * @file Automatically generated by barrelsby.
 */

export * from './SourceFacotry';
export * from './WmtsFactory';
