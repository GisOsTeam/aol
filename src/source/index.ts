/**
 * @file Automatically generated by barrelsby.
 */

export * from './ExternalVector';
export * from './Image';
export * from './ImageArcGISRest';
export * from './ImageStatic';
export * from './ImageWms';
export * from './LocalVector';
export * from './QueryArcGISRest';
export * from './TileArcGISRest';
export * from './TileImage';
export * from './TileWms';
export * from './Vector';
export * from './VectorTile';
export * from './Wfs';
export * from './Xyz';
export * from './query/index';
export * from './types/index';
