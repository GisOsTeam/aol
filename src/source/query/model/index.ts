/**
 * @file Automatically generated by barrelsby.
 */

export * from './AgsIdentifyRequest';
export * from './AgsQueryRequest';
