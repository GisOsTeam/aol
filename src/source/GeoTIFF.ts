import OlGeoTIFF, { Options } from 'ol/source/GeoTIFF';
import { ISnapshotOptions, ISnapshotSource } from './IExtended';
import { LayerType, LayerTypeEnum, SourceType, SourceTypeEnum } from './types';

export interface IGeoTIFFOptions extends ISnapshotOptions, Options {}

export class GeoTIFF extends OlGeoTIFF implements ISnapshotSource {
  protected options: IGeoTIFFOptions;

  constructor(options: IGeoTIFFOptions) {
    super({ ...options });
    this.options = { ...options };
    if (this.options.snapshotable != false) {
      this.options.snapshotable = true;
    }
    if (this.options.listable != false) {
      this.options.listable = true;
    }
    if (this.options.removable != false) {
      this.options.removable = true;
    }
  }

  public getSourceType(): SourceType {
    return SourceTypeEnum.GeoTIFF;
  }

  public getSourceOptions(): IGeoTIFFOptions {
    return this.options;
  }

  public setSourceOptions(options: IGeoTIFFOptions): void {
    this.options = { ...options };
  }

  public getLayerType(): LayerType {
    return LayerTypeEnum.WebGLTile;
  }

  public isSnapshotable(): boolean {
    return this.options.snapshotable;
  }

  public isListable(): boolean {
    return this.options.listable;
  }

  public isRemovable(): boolean {
    return this.options.removable;
  }
}
