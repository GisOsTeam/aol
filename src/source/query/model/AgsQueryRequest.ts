import { IExtended, IFeatureType, IIdentifyRequest, IQueryRequest } from '../../IExtended';
import { fromCircle } from 'ol/geom/Polygon';
import Circle from 'ol/geom/Circle';
import { FilterBuilder, FilterBuilderTypeEnum } from '../../../filter';
import EsriJSON from 'ol/format/EsriJSON';

interface AgsQueryRequestParameters {
  f: string;
  text: string;
  geometry: string;
  geometryType: string;
  inSR: string;
  spatialRel: string;
  relationParam: string;
  where: string;
  objectIds: string;
  time: string;
  distance: string;
  units: string;
  outFields: string;
  returnGeometry: string;
  maxAllowableOffset: string;
  geometryPrecision: string;
  outSR: string;
  returnIdsOnly: string;
  returnCountOnly: string;
  returnExtentOnly: string;
  orderByFields: string;
  returnZ: string;
  returnM: string;
  resultOffset: string;
  rangeValues: string;
  parameterValues: string;
  historicMoment: string;
}

export class AgsQueryRequest implements AgsQueryRequestParameters {
  f: string;
  distance: string;
  geometry: string;
  geometryPrecision: string;
  geometryType: string;
  historicMoment: string;
  inSR: string;
  maxAllowableOffset: string;
  objectIds: string;
  orderByFields: string;
  outFields: string;
  outSR: string;
  parameterValues: string;
  rangeValues: string;
  relationParam: string;
  resultOffset: string;
  returnCountOnly: string;
  returnExtentOnly: string;
  returnGeometry: string;
  returnIdsOnly: string;
  returnM: string;
  returnZ: string;
  spatialRel: string;
  text: string;
  time: string;
  units: string;
  where: string;

  private format = new EsriJSON();

  constructor(source: IExtended, type: IFeatureType<number>, request: IQueryRequest) {
    const { geometryProjection } = request;
    this.inSR = '3857';
    this.outSR = '3857';

    if ((request as IIdentifyRequest).srId) {
      this.inSR = (request as IIdentifyRequest).srId;
      this.outSR = (request as IIdentifyRequest).srId;
    }

    this.outFields = '*';
    this.returnGeometry = !!(request as IIdentifyRequest).returnGeometry ? 'true' : 'false';

    let geometry = request.geometry;
    if (geometry) {
      if (geometry.getType() === 'Circle') {
        geometry = fromCircle(geometry as Circle);
      }
      const geometryStr = this.format.writeGeometry(geometry, {
        featureProjection: geometryProjection,
        dataProjection: 'EPSG:' + this.inSR,
      });
      let geometryType = '';
      switch (geometry.getType()) {
        case 'Point':
          geometryType = 'esriGeometryPoint';
          break;
        case 'LineString':
          geometryType = 'esriGeometryPolyline';
          break;
        case 'LinearRing':
          geometryType = 'esriGeometryPolyline';
          break;
        case 'Polygon':
          geometryType = 'esriGeometryPolygon';
          break;
        case 'MultiPoint':
          geometryType = 'esriGeometryMultipoint';
          break;
        case 'MultiLineString':
          geometryType = 'esriGeometryPoint';
          break;
        case 'MultiPolygon':
          geometryType = 'esriGeometryPolygon';
          break;
      }

      this.geometry = geometryStr;
      this.geometryType = geometryType;
      this.geometryPrecision = !!(request as IIdentifyRequest).geometryPrecision
        ? `${(request as IIdentifyRequest).geometryPrecision}`
        : '10';
    }

    if (request.filters) {
      this.where = new FilterBuilder(request.filters).build(FilterBuilderTypeEnum.SQL);
    }
    this.f = 'json';
  }

  public getSrId(): string {
    return this.outSR;
  }
}
