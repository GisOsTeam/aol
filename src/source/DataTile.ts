import { GeoTIFF } from './GeoTIFF';

export type DataTile = GeoTIFF;
