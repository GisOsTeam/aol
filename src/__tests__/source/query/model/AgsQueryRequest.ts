import OlMap from 'ol/Map';
import OlView from 'ol/View';
import { ImageArcGISRest } from '../../../../source';
import { IFeatureType, IGisRequest } from '../../../../source/IExtended';
import Polygon from 'ol/geom/Polygon';
import { get as getProjection, Projection } from 'ol/proj';
import { AgsQueryRequest } from '../../../../source/query/model/AgsQueryRequest';

const states = new ImageArcGISRest({
  url: 'https://sampleserver6.arcgisonline.com/arcgis/rest/services/USA/MapServer',
  types: [{ id: 2, identifierAttribute: { key: 'objectid' } }],
});

const type: IFeatureType<number> = states.get('types')[0];

test('Simple AgsQueryRequest', () => {
  const request: IGisRequest = {
    olMap: new OlMap({
      view: new OlView({
        center: [0, 0],
        zoom: 1,
      }),
    }),
    geometry: new Polygon([
      [
        [-11580733.168194728, 4880526.966432655],
        [-11580733.168194728, 4964673.98309528],
        [-11496586.151532097, 4964673.983095286],
        [-11496586.151532097, 4880526.966432655],
        [-11580733.168194728, 4880526.966432655],
      ],
    ]),
    geometryProjection: getProjection('EPSG:3857') as Projection,
    queryType: 'query',
  };
  const agsQueryRequest = new AgsQueryRequest(states, type, request);

  expect<string>(agsQueryRequest.f).toMatchSnapshot('agsQueryRequest.f');
  expect<string>(agsQueryRequest.distance).toMatchSnapshot('agsQueryRequest.distance');
  expect<string>(agsQueryRequest.geometry).toMatchSnapshot('agsQueryRequest.geometry');
  expect<string>(agsQueryRequest.geometryPrecision).toMatchSnapshot('agsQueryRequest.geometryPrecision');
  expect<string>(agsQueryRequest.geometryType).toMatchSnapshot('agsQueryRequest.geometryType');
  expect<string>(agsQueryRequest.historicMoment).toMatchSnapshot('agsQueryRequest.historicMoment');
  expect<string>(agsQueryRequest.inSR).toMatchSnapshot('agsQueryRequest.inSR');
  expect<string>(agsQueryRequest.maxAllowableOffset).toMatchSnapshot('agsQueryRequest.maxAllowableOffset');
  expect<string>(agsQueryRequest.objectIds).toMatchSnapshot('agsQueryRequest.objectIds');
  expect<string>(agsQueryRequest.orderByFields).toMatchSnapshot('agsQueryRequest.orderByFields');
  expect<string>(agsQueryRequest.outFields).toMatchSnapshot('agsQueryRequest.outFields');
  expect<string>(agsQueryRequest.outSR).toMatchSnapshot('agsQueryRequest.outSR');
  expect<string>(agsQueryRequest.parameterValues).toMatchSnapshot('agsQueryRequest.parameterValues');
  expect<string>(agsQueryRequest.rangeValues).toMatchSnapshot('agsQueryRequest.rangeValues');
  expect<string>(agsQueryRequest.relationParam).toMatchSnapshot('agsQueryRequest.relationParam');
  expect<string>(agsQueryRequest.resultOffset).toMatchSnapshot('agsQueryRequest.resultOffset');
  expect<string>(agsQueryRequest.returnCountOnly).toMatchSnapshot('agsQueryRequest.returnCountOnly');
  expect<string>(agsQueryRequest.returnExtentOnly).toMatchSnapshot('agsQueryRequest.returnExtentOnly');
  expect<string>(agsQueryRequest.returnGeometry).toMatchSnapshot('agsQueryRequest.returnGeometry');
  expect<string>(agsQueryRequest.returnIdsOnly).toMatchSnapshot('agsQueryRequest.returnIdsOnly');
  expect<string>(agsQueryRequest.returnM).toMatchSnapshot('agsQueryRequest.returnM');
  expect<string>(agsQueryRequest.returnZ).toMatchSnapshot('agsQueryRequest.returnZ');
  expect<string>(agsQueryRequest.spatialRel).toMatchSnapshot('agsQueryRequest.spatialRel');
  expect<string>(agsQueryRequest.text).toMatchSnapshot('agsQueryRequest.text');
  expect<string>(agsQueryRequest.time).toMatchSnapshot('agsQueryRequest.time');
  expect<string>(agsQueryRequest.units).toMatchSnapshot('agsQueryRequest.units');
  expect<string>(agsQueryRequest.where).toMatchSnapshot('agsQueryRequest.where');
});
