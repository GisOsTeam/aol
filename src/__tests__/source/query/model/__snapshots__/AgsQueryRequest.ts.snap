// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Simple AgsQueryRequest: agsQueryRequest.distance 1`] = `undefined`;

exports[`Simple AgsQueryRequest: agsQueryRequest.f 1`] = `"json"`;

exports[`Simple AgsQueryRequest: agsQueryRequest.geometry 1`] = `"{"hasZ":false,"hasM":false,"rings":[[[-11580733.168194728,4880526.966432655],[-11580733.168194728,4964673.98309528],[-11496586.151532097,4964673.983095286],[-11496586.151532097,4880526.966432655],[-11580733.168194728,4880526.966432655]]]}"`;

exports[`Simple AgsQueryRequest: agsQueryRequest.geometryPrecision 1`] = `"10"`;

exports[`Simple AgsQueryRequest: agsQueryRequest.geometryType 1`] = `"esriGeometryPolygon"`;

exports[`Simple AgsQueryRequest: agsQueryRequest.historicMoment 1`] = `undefined`;

exports[`Simple AgsQueryRequest: agsQueryRequest.inSR 1`] = `"3857"`;

exports[`Simple AgsQueryRequest: agsQueryRequest.maxAllowableOffset 1`] = `undefined`;

exports[`Simple AgsQueryRequest: agsQueryRequest.objectIds 1`] = `undefined`;

exports[`Simple AgsQueryRequest: agsQueryRequest.orderByFields 1`] = `undefined`;

exports[`Simple AgsQueryRequest: agsQueryRequest.outFields 1`] = `"*"`;

exports[`Simple AgsQueryRequest: agsQueryRequest.outSR 1`] = `"3857"`;

exports[`Simple AgsQueryRequest: agsQueryRequest.parameterValues 1`] = `undefined`;

exports[`Simple AgsQueryRequest: agsQueryRequest.rangeValues 1`] = `undefined`;

exports[`Simple AgsQueryRequest: agsQueryRequest.relationParam 1`] = `undefined`;

exports[`Simple AgsQueryRequest: agsQueryRequest.resultOffset 1`] = `undefined`;

exports[`Simple AgsQueryRequest: agsQueryRequest.returnCountOnly 1`] = `"false"`;

exports[`Simple AgsQueryRequest: agsQueryRequest.returnExtentOnly 1`] = `undefined`;

exports[`Simple AgsQueryRequest: agsQueryRequest.returnGeometry 1`] = `"true"`;

exports[`Simple AgsQueryRequest: agsQueryRequest.returnIdsOnly 1`] = `"false"`;

exports[`Simple AgsQueryRequest: agsQueryRequest.returnM 1`] = `undefined`;

exports[`Simple AgsQueryRequest: agsQueryRequest.returnZ 1`] = `undefined`;

exports[`Simple AgsQueryRequest: agsQueryRequest.spatialRel 1`] = `undefined`;

exports[`Simple AgsQueryRequest: agsQueryRequest.text 1`] = `undefined`;

exports[`Simple AgsQueryRequest: agsQueryRequest.time 1`] = `undefined`;

exports[`Simple AgsQueryRequest: agsQueryRequest.units 1`] = `undefined`;

exports[`Simple AgsQueryRequest: agsQueryRequest.where 1`] = `undefined`;
